templates/edit.html

```diff
@@ -128,11 +173,100 @@ <h3 id="source-field-name" data-field_name="<%= field_name %>">
+        function getSourceInput(e){
+            // Get the raw input element for a source.
             var source_id = $(e.target).data('source_id');
             var field_name = $(e.target).data('field_name');
-            var selector = 'input[name="' + field_name + '_source"][value="' + source_id + '"]';
-            $(selector).remove();
+            var selector = 'input[class="' + field_name + '_source"][value="' + source_id + '"]';
+            return $(selector);
+        }
+
+        function removeSource(e){
+            // Remove the input element for a Source from the DOM (hard delete).
+            var $sourceInput = getSourceInput(e);
+            $sourceInput.remove();
+        }
+
+        function disableSource(e){
+            // Disable the input element for a Source, preventing it from submitting
+            // with the form (soft delete).
+            var $sourceInput = getSourceInput(e);
+            $sourceInput.removeAttr('name');  // Unnamed inputs won't be submitted by forms
+        }
+
+        function revertSource(e){
+            // Re-enable the input element for a Source.
+            var $sourceInput = getSourceInput(e);
+            $sourceInput.attr('name', $(e.target).data('field_name') + '_source');
+        }
+
+        function removeExistingSource(e){
+            // Remove a Source that has already been saved during a previous editing session.
+            disableSource(e);
+            // Alter the appearance of the alert and move it to the 'Removed' section.
+            $sourceCard = $(e.target).clone();
+            $sourceCard.removeClass('alert-success').addClass('alert-danger');
+            $sourceCard.find('.fa-remove').removeClass('fa-remove').addClass('fa-undo');
+            $sourceCard.prependTo('#source-list-removed');
+
+            // Check if we need to display or hide any section titles.
+            // These checks will be evaluated _before_ the existing source is removed
+            // from the DOM, so checking if the parent container has 1 child element
+            // will effectively check if the parent container is empty.
+            var hasNewRemovedSources = $('#source-list-removed').children().length === 1;
+            if (hasNewRemovedSources) {
+                // Show the title of the Removed Source list if this is the first source.
+                $('#source-list-removed-title').children().show();
+            }
+            var hasNoExistingSources = $('#source-list-existing').children().length === 1;
+            if (hasNoExistingSources) {
+                // Hide the title of the Existing Source list if no sources remain.
+                $('#source-list-existing-title').children().hide();
+            }
+        }
+
+        function removeAddedSource(e){
+            // Remove a Source that has been added during this editing session.
+            removeSource(e);
+            var hasNoAddedSources = $('#source-list-added').children().length === 1;
```

Comment:
Similar to the comment on line 218 above: Since we expect this function to fire before the actual alert element has been removed from the DOM, the appropriate check is to see whether the parent container has one element (the element being removed) instead of 0, as we might expect.